apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexSmartContractService.java

```diff
@@ -0,0 +1,168 @@
+        paxContractAddress = propertiesHolder.getStringProperty("apl.eth.pax.contract.address");
+        this.dexEthService = dexEthService;
+        this.ethereumWalletService = ethereumWalletService;
+    }
+
+    /**
+     *  Deposit(freeze) money(eth or pax) on the contract.
+     * @param currency Eth or Pax
+     * @return String transaction hash.
+     */
+    public String deposit(String passphrase, long accountId, String fromAddress, BigInteger weiValue, Long gas, DexCurrencies currency) throws ExecutionException, AplException.ExecutiveProcessException {
+        WalletKeysInfo keyStore = keyStoreService.getWalletKeysInfo(passphrase, accountId);
+        EthWalletKey ethWalletKey = keyStore.getEthWalletForAddress(fromAddress);
+        Long gasPrice = gas;
+
+        if(gasPrice == null){
+            gasPrice = dexEthService.getEthPriceInfo().getFastSpeedPrice();
+        }
+
+       if(gasPrice == null){
+           throw new AplException.ThirdServiceIsNotAvailable("Eth Price Info is not available.");
+       }
+        if(!currency.isEthOrPax()){
+            throw new UnsupportedOperationException("This function not supported this currency " + currency.name());
+        }
+
+        if(currency.isPax()){
+            ethereumWalletService.sendApproveTransaction(passphrase, accountId, fromAddress, smartContractAddress, weiValue);
+        }
+
+        return deposit(ethWalletKey.getCredentials(), weiValue, gasPrice, currency.isEth() ? null : paxContractAddress);
+    }
+
+
+    /**
+     *  Withdraw money(eth or pax) from the contract.
+     * @param currency Eth or Pax
+     * @return String transaction hash.
+     */
+    public String withdraw(String passphrase, long accountId, String fromAddress,  BigInteger weiValue, Long gas, DexCurrencies currency) throws AplException.ExecutiveProcessException {
+        WalletKeysInfo keyStore = keyStoreService.getWalletKeysInfo(passphrase, accountId);
+        EthWalletKey ethWalletKey = keyStore.getEthWalletForAddress(fromAddress);
+
+        if(!currency.isEthOrPax()){
+            throw new UnsupportedOperationException("This function not supported this currency " + currency.name());
+        }
+        Long gasPrice = gas;
+
+        if(gasPrice == null){
+            try {
+                gasPrice = dexEthService.getEthPriceInfo().getAverageSpeedPrice();
+            } catch (ExecutionException e) {
+                throw new AplException.ExecutiveProcessException("Third service is not available, try later.");
+            }
+        }
+
+        return withdraw(ethWalletKey.getCredentials(), weiValue, gasPrice, currency.isEth() ? null : paxContractAddress);
+    }
+
+    /**
+     * Get information how much money frozen a particular address.
+     * @return Amount of the froze money on the contract for the address.
+     */
+    public BigDecimal getUserInfoEth(String address, DexCurrencies currency){
+        if(!currency.isEthOrPax()){
```

Comment:
NPE can happen without incoming param check.

Reply:
It's impossible,  currency is required parameter.  All validation on the Controller layer.